File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        bool isbalanced;
        if (root == nullptr) {
            return true;
        }
        int leftHeight = computeNodeHeight(root->left);
        int rightHeight = computeNodeHeight(root->right);
        if (std::abs(leftHeight - rightHeight) > 1) {
            return false;
        }
        return isBalanced(root->left) && isBalanced(root->right);

    }
private: 
    // Function to compute the height of a given node
    int computeNodeHeight(TreeNode* node) {
        // Base case: If the node is null, its height is defined as -1
        if (node == nullptr) {
            return -1;
        }

        // Recursively calculate the height of the left and right subtrees
        int leftHeight = computeNodeHeight(node->left);
        int rightHeight = computeNodeHeight(node->right);

        // The height of the current node is the maximum of its children's heights plus one
        return std::max(leftHeight, rightHeight) + 1;
    }
};

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        counts = {}
        maxCount = left = maxLength = 0

        for right in range(len(s)):
            counts[s[right]] = counts.get(s[right], 0) + 1
            maxCount = max(maxCount, counts[s[right]])

            while (right - left + 1) - maxCount > k:
                counts[s[left]] -= 1
                left += 1

            maxLength = max(maxLength, right - left + 1)
        return maxLength
        